### ISSUE TITLE
default_sort fields should be added as indexes

### ISSUE THREAD

**Issue by sunnysideup** [silverstripe/silverstripe-framework#7272]
Sorting on an indexed field is faster: https://dev.mysql.com/doc/refman/5.7/en/order-by-optimization.html#order-by-index-use. 

Therefore, I would like to suggest that, by default, database fields listed in `default_sort` should be added as indexes. 

Not sure if it would work for `default_sorts` with multiple fields, but, for example, for `SiteTree` we should definitely add an Index on the `Sort` Field.

**Comment by tractorcow** [comment]
👍  for this, it wouldn't be too hard, and would add more value than the likely cost.

We would need to ensure that this could be disabled as necessary though. I suggest only adding this as default if an `indexes` config wasn't specified for that model.

**Comment by sunnysideup** [comment]
> We would need to ensure that this could be disabled as necessary though. I suggest only adding this as default if an indexes config wasn't specified for that model.

Agreed that you should be able to turn it off, but don't turn  if off just because indexes are set. You need both your custom indexes (for searching, etc... we often set those) and the indexes for Sort. 

Check this out (will need to be peer reviewed - I may have screwed it up):

```php
//A: sort by SiteTree.ID
$objects = SiteTree:get()->sort(['ID' => 'ASC'])->limit(10, 20);
foreach($objects as $obj) {
    echo $obj->ID;
}
//B: sort by SiteTree.Sort
$objects = SiteTree:get()->sort(['Sort' => 'ASC'])->limit(10, 20);
foreach($objects as $obj) {
    echo $obj->ID;
}
```
My results: 
SiteTree.Count = 126 => ~5% faster (highly variable from -5% through to 20%) 
SiteTree.Count = 126 => 20% faster
SiteTree.Count = 800 => 40% faster
SiteTree.Count = 6000 => 70% faster
SiteTree.Count = 40,000 => 90% faster

I use the following tool:
https://github.com/sunnysideup/silverstripe-dataintegritytests/blob/master/code/CheckForMysqlPaginationIssuesBuildTask.php#L282-L347

**Comment by sunnysideup** [comment]
After adding an index to SiteTree.Sort in a database with 40,000 records, there is no difference between `ORDER BY ID`, `ORDER BY Sort` and no `ORDER BY` .... (down from 90% faster).

We also saw a reduction to negligible in the SiteTree.Count = 800 site.

**Comment by tractorcow** [comment]
Could be done in 4.2.0

**Comment by dhensby** [comment]
fixed in #7559
